src/Products/PluginIndexes/KeywordIndex/KeywordIndex.py

```diff
@@ -58,59 +72,120 @@ def _index_object(self, documentId, obj, threshold=None, attr=''):
         # we'll do so.
 
         newKeywords = self._get_object_keywords(obj, attr)
+        oldKeywords = self._unindex.get(documentId, _marker)
 
-        oldKeywords = self._unindex.get(documentId, None)
-
-        if oldKeywords is None:
+        if oldKeywords is _marker:
             # we've got a new document, let's not futz around.
-            try:
+            if newKeywords in (missing, empty):
+                self.insertSpecialIndexEntry(newKeywords, documentId)
+            else:
+                keys = list()
                 for kw in newKeywords:
-                    self.insertForwardIndexEntry(kw, documentId)
-                if newKeywords:
-                    self._unindex[documentId] = list(newKeywords)
-            except TypeError:
-                return 0
+                    try:
+                        self.insertForwardIndexEntry(kw, documentId)
+                        keys.append(kw)
+                    except TypeError:
```

Comment:
I doubt that this exception handling is right: it does not index the object if one key cannot be indexed - and the problem is only reported via a log entry.
In my opinion, other alternatives would be better:
* log then ignore keys not indexable
* do not catch the exception (and let the whole operation fail)
* handle this `TypeError` in the same way as if it had occurred during the object value determination (e.g. map to `missing`).

In any case, the logic is at the wrong place. One would need similar logic for "update existing index info" and it should not be duplicated.

Reply:
In the original version there was a bug that could lead to inconsistencies in the index. Also, the problem was not logged. In order not to have to abolish the old behavior completely, I would prefer the first variant. Consequently, `_unindex` is only allowed to store indexable keywords.

Reply:
**Correction**: Since the type OOSet is forced for keywords in the meantime, a TypeError can also be raised under python3 e.g. in the method `map_value`. For consistency reasons, TypeError is now always handled in the same way when determining the attribute value.

Reply:
@d-maurer I'm beginning to wonder if it wouldn't be more sensible to escalate TypeError when a value in the keyword list is incompatible with the already indexed values. Otherwise the new values would have to be pre-validated before being indexed.